curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/CuratorCacheListenerBuilder.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.curator.framework.recipes.cache;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.CuratorCacheListener.Type;
+import java.util.function.Consumer;
+
+public interface CuratorCacheListenerBuilder
+{
+    /**
+     * Add a standard listener
+     *
+     * @param listener listener to add
+     * @return this
+     */
+    CuratorCacheListenerBuilder forAll(CuratorCacheListener listener);
+
+    /**
+     * Add a listener only for {@link Type#NODE_CREATED}
+     *
+     * @param listener listener to add
+     * @return this
+     */
+    CuratorCacheListenerBuilder forCreates(Consumer<ChildData> listener);
+
+    @FunctionalInterface
+    interface ChangeListener
+    {
+        void event(ChildData oldNode, ChildData node);
+    }
+
+    /**
+     * Add a listener only for {@link Type#NODE_CHANGED}
+     *
+     * @param listener listener to add
+     * @return this
+     */
+    CuratorCacheListenerBuilder forChanges(ChangeListener listener);
+
+    /**
+     * Add a listener only both {@link Type#NODE_CREATED} and {@link Type#NODE_CHANGED}
+     *
+     * @param listener listener to add
+     * @return this
+     */
+    CuratorCacheListenerBuilder forCreatesAndChanges(ChangeListener listener);
```

Comment:
Instead of having these multiple combinations (and the user will still need to check type for each event e.g. `type == NODE_CHANGED`) you could introduce interface that allows the user implement only the methods that are interesting for him/her, and have the burden of type -> action mapping specified in one location for all users:
`public interface CacheListener { 
  void onCacheInitialized();
  void onNodeCreate(ChildData data);
  void onNodeChanged(ChildData oldData, ChildData data);
  void onNodeDeleted(ChildData data);
}`
(you could introduce empty implementation for this interface for convenience).
It is approach that doesn't embrace lambdas, but in this case of multiple event types it might be more comfortable for users.

Reply:
We could have both actually. We can add a `CacheListenerAdaptor` that looks like what you have. However, since Java 8, haven't most devs moved to the neo-functional style? The benefit of how I did the listener builder is that you don't have to use cumbersome anonymous classes. i.e. you can do:

```java
CuratorCacheListener.builder()
    .forDeletes(node -> handleDelete(node))
    .forInitialized(() -> handleInitialized())
    ... etc ...
```

This is superior to anonymous inner classes.

Reply:
The adaptor would look like this. Should I add this? Is it gilding the Lilly? https://gist.github.com/Randgalt/cbe38cb3cdf1825f73dd165d997a2248

Reply:
No, I take it back. Didn't understand fully your builder until now.